[InstCombine] Fix for folding `select` into floating point binary operators.

michele-scandale opened the pull request:
Folding a `select` into a floating point binary operators can only be done if the result is preserved for both case. In particular, if the other operand of the `select` can be a NaN, then the transformation won't preserve the result value.

michele-scandale commented:
I originally filed https://github.com/llvm/llvm-project/issues/82726 and in there https://github.com/llvm/llvm-project/issues/74297 was mentioned which is basically the same issue.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -527,8 +527,11 @@ Instruction *InstCombinerImpl::foldSelectIntoOp(SelectInst &SI, Value *TrueVal,
-    if (isa<FPMathOperator>(&SI))
+    if (isa<FPMathOperator>(&SI)) {
       FMF = SI.getFastMathFlags();
+      if (!computeKnownFPClass(FalseVal, FMF, fcNan, &SI).isKnownNeverNaN())
```
Think it be better to place this below after line 543 to avoid relatively expensive analysis until after all the basic checks have been performed.

michele-scandale commented on llvm/test/Transforms/InstCombine/select-binop-foldable-floating-point.ll:
```diff
@@ -1,8 +1,19 @@
+  ret float %D
+}
+
+define float @select_fpclass_fadd(i1 %cond, float nofpclass(nan) %A, float %B) {
```
@arsenm This is the test I added to cover the case where the fast-math flags on the `select` are not present, but we still perform the optimization because we can figure out that `%A` cannot be a NaN due the `nofpclass(nan)` attribute on the function parameter.

michele-scandale commented on llvm/test/Transforms/InstCombine/select-binop-foldable-floating-point.ll:
```diff
@@ -56,202 +78,202 @@ define <4 x float> @select_nsz_fadd_v4f32(<4 x i1> %cond, <4 x float> %A, <4 x f
-  %D = select i1 %cond, float %C, float %A
+  %D = select nnan i1 %cond, float %C, float %A
   ret float %D
 }
```
See https://github.com/llvm/llvm-project/pull/83200#discussion_r1508521767

michele-scandale commented:
Ping

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -536,19 +536,24 @@ Instruction *InstCombinerImpl::foldSelectIntoOp(SelectInst &SI, Value *TrueVal,
+        (!OOpIsAPInt || !isSelect01(C->getUniqueInteger(), *OOpC)))
+      return nullptr;
+
+    // Cannot fold if the false value might be a NaN.
```
Maybe explain why

michele-scandale replied:
Updated the comment. I'll wait few days before landing the change in case there are further comments. Thanks.

michele-scandale committed:
[InstCombine] Fix for folding `select` into floating point binary operators.

Folding a `select` into a floating point binary operators can only be
done if the result is preserved for both case. In particular, if the
other operand of the `select` can be a NaN, then the transformation
won't preserve the result value.

dtcxzyw commented:
Ping. Is it ready to land?

alexfh commented:
We've noticed a 10-20% regression of the https://github.com/llvm/llvm-test-suite/tree/main/MultiSource/Benchmarks/Ptrdist/ks benchmark on AArch64 after this commit. We don't consider this an important microbenchmark, but I wanted to let you know just in case.

pawosm-arm commented:
I've found that the change introduced here has come with a huge performance penalty at `-Ofast`. Considering that `-Ofast` induces `-fno-honor-nans`, shouldn't this change be restricted to `-fhonor-nans`?

dtcxzyw commented:
> I've found that the change introduced here has come with a huge performance penalty at `-Ofast`. Considering that `-Ofast` induces `-fno-honor-nans`, shouldn't this change be restricted to `-fhonor-nans`?

`-Ofast` was deprecated in Clang 19. Can you try with `-ffast-math`?

pawosm-arm commented:
Yes, I've tried `-ffast-math` too, with the same result.

pawosm-arm commented:
Seems all we need is this change:

```
diff --git a/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp b/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
index aaf4ece3249a..c8d9b4ee446e 100644
--- a/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
+++ b/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
@@ -551,7 +551,7 @@ Instruction *InstCombinerImpl::foldSelectIntoOp(SelectInst &SI, Value *TrueVal,
     // This makes the transformation incorrect since the original program would
     // have preserved the exact NaN bit-pattern.
     // Avoid the folding if the false value might be a NaN.
-    if (isa<FPMathOperator>(&SI) &&
+    if (isa<FPMathOperator>(&SI) && SI.getFastMathFlags().noNaNs() &&
         !computeKnownFPClass(FalseVal, FMF, fcNan, &SI).isKnownNeverNaN())
       return nullptr;

```

pawosm-arm commented:
@dtcxzyw could you tell me if there is any followup planned for this?

arsenm commented:
> Seems all we need is this change:
> 
> ```
> diff --git a/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp b/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
> index aaf4ece3249a..c8d9b4ee446e 100644
> --- a/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
> +++ b/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
> @@ -551,7 +551,7 @@ Instruction *InstCombinerImpl::foldSelectIntoOp(SelectInst &SI, Value *TrueVal,
>      // This makes the transformation incorrect since the original program would
>      // have preserved the exact NaN bit-pattern.
>      // Avoid the folding if the false value might be a NaN.
> -    if (isa<FPMathOperator>(&SI) &&
> +    if (isa<FPMathOperator>(&SI) && !TVI->getFastMathFlags().noNaNs()  &&
>          !computeKnownFPClass(FalseVal, FMF, fcNan, &SI).isKnownNeverNaN())
>        return nullptr;
> ```
> 
> I had been lost in that logic at first, then I've found that it's the `TVI` that has the `-f(no-)honor-nans` flag right (with its default value for `-ffast-math` set accordingly).

Testcase for this?

pawosm-arm commented:
> Testcase for this?

@arsenm I'm using the following code for testing this: https://github.com/dslarm/Financial-Services-Workload-Samples/blob/main/MonteCarloEuropeanOptions/MonteCarloInsideBlockingDP.cpp

With the commit introduced by this PR, the reduction loop cannot be vectorized:

```
MonteCarloInsideBlockingDP.cpp:258:5: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
  258 |     #pragma omp simd reduction(+:v0) reduction(+:v1)
      |     ^
1 warning generated.
```

Normally, only a slight (yet consistent) change could be observed from that, unless veclib is used, in which case lack of the vectorization results in a dramatic performance degradation.

arsenm commented:
> > Testcase for this?
> 
> @arsenm I'm using the following code (compiled with `-fopenmp`) for testing this: https://github.com/dslarm/Financial-Services-Workload-Samples/blob/main/MonteCarloEuropeanOptions/MonteCarloInsideBlockingDP.cpp

Can you turn this into a reduced IR test, in a PR with your proposed patch above

pawosm-arm commented:
I'd need to know if such change makes sense before working on a PR with a proper test case.  Namely, is obtaining information from `TVI` the right thing to do when `SI` doesn't seem to have it. It could be the state of noNaNs flag has been deliberately stripped from `SI`, and reading it from `TVI` is just sweeping whatever happened to `SI` under the carpet.

arsenm commented:
> I'd need to know if such change makes sense before working on a PR with a proper test case. 

You need a proper test case to make sense of the situation. It would be easier to reason about and discuss with a concrete 3 line IR testcase than this 

> Namely, is obtaining information from `TVI` the right thing to do when `SI` doesn't seem to have it.

computeKnownFPClass knows a few tricks operating directly on the select based on the compared values. It would help to see the reduced example

pawosm-arm commented:
> It would be easier to reason about and discuss with a concrete 3 line IR testcase than this

Ok, so here it is:

```
  %cmp1 = fcmp fast ogt double %sub, 0.000000e+00
  %add1 = fadd fast double %v1, %sub
  %v1.1 = select nsz i1 %cmp1, double %add1, double %v1
```

without this commit, it is transformed into:

```
  %cmp1 = fcmp fast ogt double %sub, 0.000000e+00
  %add1 = select nsz i1 %cmp1, double %sub, double 0.000000e+00
  %v1.1 = fadd fast double %add1, %v1
```

giving it some more context:

```
--- fun1.ll     2024-10-22 15:21:01.291767219 +0100
+++ fun1.opt.ll 2024-10-22 15:24:15.400736173 +0100
@@ -1,12 +1,12 @@
 target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32"
 target triple = "aarch64-unknown-linux-gnu"

 @v0 = global double 0.000000e+00

 define void @fun(double %v1, double %sub) {
   %cmp1 = fcmp fast ogt double %sub, 0.000000e+00
-  %add1 = fadd fast double %v1, %sub
-  %v1.1 = select nsz i1 %cmp1, double %add1, double %v1
+  %add1 = select nsz i1 %cmp1, double %sub, double 0.000000e+00
+  %v1.1 = fadd fast double %add1, %v1
   store double %v1.1, ptr @v0, align 8
   ret void
 }
```

pawosm-arm commented:
If I add the `fast` attribute to the `select` in my example code:
```
  %cmp1 = fcmp fast ogt double %sub, 0.000000e+00
  %add1 = select nsz i1 %cmp1, double %sub, double 0.000000e+00
  %v1.1 = fadd fast double %add1, %v1
```
the transformation succeed even with your patch. It's because of this:
```
    // TODO: We probably ought to revisit cases where the select and FP
    // instructions have different flags and add tests to ensure the
    // behaviour is correct.
    FastMathFlags FMF;
    if (isa<FPMathOperator>(&SI))
      FMF = SI.getFastMathFlags();
```
`select` isn't fast but other things are, so we hit the `TODO:` situation.

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -536,19 +536,29 @@ Instruction *InstCombinerImpl::foldSelectIntoOp(SelectInst &SI, Value *TrueVal,
+    // have preserved the exact NaN bit-pattern.
+    // Avoid the folding if the false value might be a NaN.
+    if (isa<FPMathOperator>(&SI) &&
+        !computeKnownFPClass(FalseVal, FMF, fcNan, &SI).isKnownNeverNaN())
```
This ought to try to make use of the nnan flag on the select before trying the computeKnownFPClass query. There's a wrapper around computeKnownFPClass which takes an additional FastMathFlags

pawosm-arm replied:
I'm sorry, I'm sometimes lost here.
`There's a wrapper around computeKnownFPClass which takes an additional FastMathFlags`
Did you mean
`The wrapper around computeKnownFPClass called here takes an additional FastMathFlags`? That's how I see here, InstCombineSelect.cpp defines two wrappers around `computeKnownFPClass`:
```
  KnownFPClass computeKnownFPClass(Value *Val, FastMathFlags FMF,
                                   FPClassTest Interested = fcAllFlags,
                                   const Instruction *CtxI = nullptr,
                                   unsigned Depth = 0) const
```
and
```
  KnownFPClass computeKnownFPClass(Value *Val,
                                   FPClassTest Interested = fcAllFlags,
                                   const Instruction *CtxI = nullptr,
                                   unsigned Depth = 0) const
```
The one called here is the one which takes `FastMathFlags FMF`.

pawosm-arm replied:
> This ought to try to make use of the nnan flag on the select before trying the computeKnownFPClass query

Could you say what use would that be?

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -536,19 +536,29 @@ Instruction *InstCombinerImpl::foldSelectIntoOp(SelectInst &SI, Value *TrueVal,
+    NewSel->takeName(TVI);
+    BinaryOperator *BO =
+        BinaryOperator::Create(TVI->getOpcode(), FalseVal, NewSel);
+    BO->copyIRFlags(TVI);
```
I think this can or in the value flags from the select

arsenm replied:
alive2 is really slow here without known-no-infs on the inputs though, so I'm not sure the exact flag rewrite rules yet

arsenm replied:
maybe (nnan_bo | nnan_select) | (others_flags_bo & other_flags_select)

pawosm-arm replied:
> I think this can or in the value flags from the select

I don't even know what does it mean :(

pawosm-arm replied:
> > I think this can or in the value flags from the select
> 
> I don't even know what does it mean :(

Ok, I guess I've deciphered it. I guess what you propose is to 'or' the flags copied from TVI with value flags from the select. Trouble is, SI don't have any flags, they were not generated by simplifyCFG pass since they were lost in SROA pass.
